### PR TITLE
IO.warn to unify warnings

### PR DESCRIPTION
**Pull request by lpil**
How does this look for #4608 ?

### CODE REVIEW

**Comment by josevalim** [comment]
Thank you @lpil! Let's please make the stacktrace (the stacktrace and not the result of Exception.format_stacktrace) an optional second argument. So we can call it like this:

    IO.warn "foo bar"

Notice you will have to implement the stacktrace gathering function like this:

    def warn(message, trace \\ nil) do
      trace = trace || case Process.info(self, :current_stacktrace) do
        {:current_stacktrace, t} -> Enum.drop(t, 3)
      end
      ...
    end

And then call Exception.format_stacktrace to format it. Check Exception.format_stacktrace itself for an example. :)

**Comment by lpil** [comment]
Sounds great. :)

Is there a mechanism for printing a warning without the stack trace?

**Comment by josevalim** [comment]
@lpil you will pass an empty list. :)

**Comment by lpil** [comment]
Makes sense. I'll document that. Thanks :)

File: lib/elixir/test/elixir/io_test.exs [hunk 1]
```diff
@@ -118,6 +118,19 @@ defmodule IOTest do
     assert capture_io(fn -> IO.puts(13) end) == "13\n"
   end
 
+  test "warn with chardata" do
+    assert capture_io(:stderr, fn -> IO.warn("hello") end) =~ "warning: hello\n    (ex_unit) lib/ex_unit"
+    assert capture_io(:stderr, fn -> IO.warn('hello') end) =~ "warning: hello\n    (ex_unit) lib/ex_unit"
+    assert capture_io(:stderr, fn -> IO.warn(:hello) end) =~ "warning: hello\n    (ex_unit) lib/ex_unit"
+    assert capture_io(:stderr, fn -> IO.warn(13) end) =~ "warning: 13\n    (ex_unit) lib/ex_unit"
+    assert capture_io(:stderr, fn -> IO.warn("hello", []) end) == "warning: hello\n\n"
```
Comment (lpil) [inline lib/elixir/test/elixir/io_test.exs]: @josevalim What do you think about this double newline when passing an empty list?

Comment (josevalim) [inline lib/elixir/test/elixir/io_test.exs]: @lpil we should probably get rid of it otherwise someone will certainly report it as an error :)

**Comment by lpil** [comment]
OK, no more double newline. How does that look @josevalim ?

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

**Comment by lpil** [comment]
Thanks! :)

**Pull request merged**